src/librustc_typeck/coherence/overlap.rs

```diff
@@ -48,28 +48,28 @@ impl<'cx, 'tcx> OverlapChecker<'cx, 'tcx> {
             Value,
         }
 
-        let name_and_namespace = |def_id| {
-            let item = self.tcx.associated_item(def_id);
+        let name_and_namespace = |r: &ty::AssociatedItemRef| {
+            let item = self.tcx.associated_item(r.def_id);
```

Comment:
This is one of the cases where it would make sense to also store the kind of associated item in the ref.